src/NextjsDomain.ts

```diff
@@ -0,0 +1,145 @@
+import { ICertificate, Certificate, CertificateValidation } from 'aws-cdk-lib/aws-certificatemanager';
+import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
+import {
+  ARecord,
+  ARecordProps,
+  AaaaRecord,
+  AaaaRecordProps,
+  HostedZone,
+  IHostedZone,
+  RecordTarget,
+} from 'aws-cdk-lib/aws-route53';
+import { HttpsRedirect } from 'aws-cdk-lib/aws-route53-patterns';
+import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
+import { Construct } from 'constructs';
+
+export interface NextjsDomainProps {
+  /**
+   * The domain to be assigned to the website URL (ie. domain.com).
+   * Supports domains that are hosted either on [Route 53](https://aws.amazon.com/route53/) or externally.
+   */
+  readonly domainName: string;
+  /**
+   * An alternative domain to be assigned to the website URL. Visitors to the alias will be redirected to the main domain. (ie. `www.domain.com`).
+   * Use this to create a `www.` version of your domain and redirect visitors to the root domain.
+   */
+  readonly domainAliases?: string[];
+  /**
+   * Specify additional names that should route to the Cloudfront Distribution. Note, certificates for these names will not be automatically generated so the `certificate` option must be specified.
+   */
+  readonly alternateNames?: string[];
+  /**
+   * Set this option if the domain is not hosted on Amazon Route 53.
```

Comment:
```suggestion
   * Set this option if the domain is not hosted on Amazon Route 53 or is hosted on Route53 but in a different account or otherwise in a way it is unaccessible. If false, the domain records will not be automatically created.
```

Maybe this is wordy, and may not totally clear, but this was something that threw me off. a bit. I think the name is still fine ("external domain"), but when I originally read this I thought "oh, no, my domain is in AWS" so set to true... but really it only matters if the account / IAM role running the CDK / CloudFormation is able to create Route53 records on behalf of this domain.

Reply:
@kevin-mitchell, agree with you it's not clear. Now is the time if we're going to change the name of this variable. Do you think there is a better alternative?

Reply:
One option could be to remove `isExternalDomain` and require that user runs
```ts
HostedZone.fromLookup(this, 'HostedZone', {
          domainName: this.props.domainName,
        });
```
from user-land?

Reply:
@kevin-mitchell, I appreciate your thoughts. If user supplies `isExternalDomain`, then `hostedZone` isn't created, which means `Certificate` isn't created and DNS records aren't created. So it seems to me like there isn't much value to `NextjsDomain` if `isExternalDomain: true`. Do you agree?

Therefore, I think it makes sense to remove the `isExternalDomain` prop and require `hostedZone`. What do you think?

Reply:
A potential issue with requiring `hostedZone`, is that if user wants/has to use another DNS provider, then how do they configure CloudFront Distribution to use that DNS? I guess they could use `defaults` to customize props passed into CloudFront Distribution? For example, if they want to provide their own `Certificate` that they created and validated themselves by adding DNS records to namecheap?

Reply:
@bestickley 

I agree there isn't _much_ value in `isExternalDomain`, and at worse it can caused confusion. I'm sort of repeating what I said above but to me it seems like either somebody is sufficiently advance to be able to intuit what needs to be done for their specific use case, or setting `isExternalDomain` isn't really going to magically make everything work anyway. So from that perspective there isn't much value IMHO.

BUT, it is at least a "clue" I suppose, if somebody sets it to true they likely will also read the comment and that might give us an opportunity to explain a bit more about what the flag does, and what we'd expect them to take care of on their side if their domain really is external (e.g. manually create A/AAAA records for the CloudFront distribution, make sure there is a valid certificate, etc).

**BUT I don't think simply removing it is really an option actually**, as it would make some things impossible (this would actually break things for me)** This code would not work for me, because in my case I `isExternalDomain` as "true", BUT I pass in a certificate. If we simply removed the prop, then in my case the HostedZone.fromLookup would run `return HostedZone.fromLookup(this, 'Ho...`, which would fail. 

So I'd either keep it for now I think (it's late here in Japan so I might not be thinking perfectly, I can take another look in my morning tomorrow or of course I defer to you :))


```
  private getHostedZone(): IHostedZone | undefined {
    if (this.props.isExternalDomain) {
      return;
    } else {
      if (!this.props.hostedZone) {
        return HostedZone.fromLookup(this, 'HostedZone', {
          domainName: this.props.domainName,
        });
      } else {
        return this.props.hostedZone;
      }
    }
  }

  private getCertificate(): ICertificate | undefined {
    if (this.props.certificate) {
      return this.props.certificate;
    } else if (this.hostedZone) {
      return new Certificate(this, 'Certificate', {
        domainName: this.props.domainName,
        validation: CertificateValidation.fromDns(this.hostedZone),
      });
    } else {
      return;
    }
  }
```

Reply:
@kevin-mitchell, great point. Can you send me the props you're passing with this new version?
If I'm understanding correctly, you're passing in certificate and hostedZone and having `NextjsDomain` create DNS records for you?

Reply:
Because if you're not passing `hostedZone` and only passing `certificate`, then you could just pass the certificate to default CloudFront Distribution props to get the certificate to distribution and bypass `NextjsDomain`, right?

Reply:
Just to add to the above in case the context is useful to anybody in the future reading this, what I do is create the certificate myself outside of the Constructs in this project, in `us-east-1` - apparently required. see: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cnames-and-https-requirements.html but it says

> We recommend that you use a certificate issued by [AWS Certificate Manager (ACM)](https://aws.amazon.com/certificate-manager/). For information about getting a certificate from ACM, see the [AWS Certificate Manager User Guide](https://docs.aws.amazon.com/acm/latest/userguide/). To use an ACM certificate with CloudFront, **make sure you request (or import) the certificate in the US East (N. Virginia) Region (us-east-1)**.

Note that I use `CertificateValidation.fromDns()` if I am not creating a hosted zone in my account. This is the case in my "prod" accounts, because I use DNS Zone Delegation to slice off sub-domains for my other accounts (e.g. dev.domain.com), but in the case of prod my root AWS organization account owns the hosted zone so I can't (easily) add DNS records to a different account (I think there actually might be SOME way, but I'd prefer to keep things simple and for prod I don't mind the small one time setup). 

```
    // Note that we will need to manually create the CNAME records in the root / domain account, certainly
    // for "prod" because the hosted zone won't have permission to add the records in another root account.
    const certificate = new acm.Certificate(
      this,
      `${props.prefix}-domain-certificate`,
      {
        domainName: hostedZoneDomain,
        subjectAlternativeNames:
          props.domainConfig.alternateDomains.length > 0
            ? props.domainConfig.alternateDomains
            : undefined,
        validation: customDomainHostedZone
          ? CertificateValidation.fromDns(customDomainHostedZone)
          : CertificateValidation.fromDns(),
      },
    );

    new ssm.StringParameter(this, "domain-certificate-arn", {
      parameterName: `/${props.prefix}/domain/domain-certificate-arn`,
      stringValue: certificate.certificateArn,
    });
    ```

After I run the above the deployment actually stalls because I have to go and manually add the DNS records to validate the Certificate (basically add a few CNAME records into my root domain). I do this by logging into current / child AWS console and viewing the certificate in ACM, copying the values to my root account hosted zone.

Then within the actualy NextJS construct I do this:

```
const nextjs = new Nextjs(this, `${props.prefix}-web`, {
      nextjsPath: "./apps/web",
      buildCommand: "npx open-next@2.2.4 build",
      environment: {
        PARAMETERS_SECRETS_EXTENSION_LOG_LEVEL: "warn",
        ENVIRONMENT_PREFIX: props.prefix,
       ...
      },
      domainProps: {
        domainName: props.domainName,
        // this is the certificate I generated above
        certificate: domainCert,
        alternateNames: props.alternateDomainNames,
        // Setting this to true means the NextJs construct won't attempt to
        // create a hosted zone, and if we didn't configure a hosted zone
        // e.g. because prod domains are owned by root account then we'll
        // set this to true, because the hosted zone won't be findable
        // in the AWS account
        isExternalDomain: true,
      },
      defaults: {
        lambda: {
          paramsAndSecrets: paramsAndSecretsLayerVersion,
        },
        distribution: {
          domainNames: [
            "domain.io",
            "www.domain.io",
            "prod.domain.io",
          ],
        },
      },
    });

```

Reply:
@kevin-mitchell, so I think with your code (thanks for sharing by the way), `NextjsDomain` is creating nothing for you. The only "value" it's adding for you is passing your certificate and domain names to the CloudFront Distribution which you could alternatively do with:

```ts
defaults: {
  distribution: {
    certificate: domainCert,
    domainNames: props.alternateDomainNames,
  },
}
```
Could you try? I know it's late for you, so no rush.

Reply:
I'll look in the AM, but yes, for my _prod_ environment `NextjsDomain` does nothing for me. To be clear I do use it fully for my lower environments (because domain delegation lets my lower accounts control a slice of the domain record).

I can do what you suggest but it will fail if I don't have something set in domainProps, because it also gets the domain names from the NextjsDomain:

https://github.com/jetbridge/cdk-nextjs/blob/7772bb38ce844cc1fa57d7b8fde4a4c0dfeb1be2/src/NextjsDistribution.ts#L386

https://github.com/jetbridge/cdk-nextjs/blob/7772bb38ce844cc1fa57d7b8fde4a4c0dfeb1be2/src/NextjsDomain.ts#L52

So what you suggest might work for certificate (honestly I missed this, so thank you!) but we still rely on `NextjsDomain` to pass things through. Hopefully that makes sense.

Reply:
> I can do what you suggest but it will fail if I don't have something set in domainProps, because it also gets the domain names from the NextjsDomain:

You're right, it would fail. I'm going to push a commit fixes the props interface indicating it's optional which will give me type error. Then I'll use optional chaining (`this.props.nextDomain?.domainNames`) so it doesn't break.

>  but we still rely on NextjsDomain to pass things through

I think we should change this.

Reply:
I'm leaning towards removing `isExternalDomain` because it is confusing to me to have a single prop in a construct make it so that no constructs are instantiated in that construct. I could be convinced otherwise.
@revmischa, any thoughts?

Reply:
> I have changed this.

FYI, maybe you didn't push yet, but I believe as of now the `nextDomain` is still where the certificate and domain names are being pulled from https://github.com/jetbridge/cdk-nextjs/pull/174/files#diff-1c29a62fdd463658d0b7511d7e6ff52f8d99a7bdb242659b103536cc5e05d63bR386-R387

Reply:
@kevin-mitchell, I didn't see that `...cfDistributionProps,` was before `domainNames` and `certificate` in the object so you currently cannot override. I just fixed that.

Reply:
would a prop like createRoute53Records: boolean be clearer?

Reply:
@revmischa, good suggestion. `createRoute53Records` is more a explicit/procedural name. However currently I'm leaning towards removing it because I think it's a little strange/misleading/confusing to have a prop then makes it so no child constructs are created in the construct you're instantiating. Do you agree?

Reply:
I'm a fan of making the typical happy path simple. Most people who deploy a nextjs site with a custom domain on AWS probably would like the route53 records created for them. They could be created in a separate construct if this one is getting too overloaded.

Reply:
> would like the route53 records created for them

I completely agree! And `NextjsDomain` will do this for the user. What I'm suggesting is not to make it more difficult for the typical happy path, but more clear for the complex use case by removing `isExternalDomain` which can and has caused confusion to users. This will tell users they need to use their own constructs outside `Nextjs`